### ISSUE TITLE
Incorrect suggestion when contains_key+insert is used in a while loop

### ISSUE THREAD

**Issue by aditya-r-m** [rust-lang/rust-clippy#10331]
### Summary

When contains_key+insert is used in a while loop, Clippy suggestion is an unnecessary unrolling of the code.

### Lint Name

map_entry

### Reproducer

Original Code (Expected):

```rust
    while !fi.contains_key(&x) {
        fi.insert(x, i);
        i += 1;
        x = fx(x);
    }
```

Clippy Suggestion:

```rust
     if let std::collections::hash_map::Entry::Vacant(e) = fi.entry(x) {
         e.insert(i);
         i += 1;
         x = fx(x);
     } else while !fi.contains_key(&x) {
         fi.insert(x, i);
         i += 1;
         x = fx(x);
     }
```

### Version

```text
rustc 1.67.1 (d5a82bbd2 2023-02-07)
binary: rustc
commit-hash: d5a82bbd26e1ad8b7401f6a718a9c57c96905483
commit-date: 2023-02-07
host: x86_64-unknown-linux-gnu
release: 1.67.1
LLVM version: 15.0.6
```
